--- src/jdk.hotspot.agent/solaris/native/libsaproc/saproc.cpp~	Thu Jan 30 10:08:30 2020
+++ src/jdk.hotspot.agent/solaris/native/libsaproc/saproc.cpp	Wed Apr 15 15:14:53 2020
@@ -27,7 +27,11 @@
 #include <thread_db.h>
 #include <strings.h>
 #include <limits.h>
+#ifdef __GNUC__
+#include <cxxabi.h>
+#else
 #include <demangle.h>
+#endif
 #include <stdarg.h>
 #include <stdlib.h>
 #include <errno.h>
@@ -1239,8 +1239,19 @@
   CHECK_EXCEPTION_(NULL);
   char  buf[2*SYMBOL_BUF_SIZE + 1];
   jstring res = 0;
+#ifdef __GNUC__
+  int   status;
+  char* result;
+  if ((result = abi::__cxa_demangle(ptr, NULL, NULL, &status)) != NULL) {
+    snprintf(buf, 2*SYMBOL_BUF_SIZE+1, "%s", result);
+#else
   if (cplus_demangle((char*) ptr, buf, sizeof(buf)) != DEMANGLE_ESPACE) {
+#endif
     res = env->NewStringUTF(buf);
+#ifdef __GNUC__
+    // call c library's free
+    ::free(result);
+#endif
   } else {
     res = name;
   }
